### ISSUE TITLE
[Question] Accessing particular models found by autosklearn

### ISSUE THREAD

**Issue by jppmatos** [automl/auto-sklearn#1252]
# How to properly save and load a pipeline/model?

Hi, I want to predict data, using new test data, with the found model.

I could use pickle to save and load the found model, but does not include the feature pre-processing (_data_preprocessing_)?  (issue #849 and **manual - Model persistence**)
Do I must use `fit_pipeline()` , or something else, just to transform data before predicting?

If it is just using pickle, I just need to do this: 
```
#...

#preform the seach:
model_autoSK.fit(X_train,y_train)

# save model
with open('model-classifier.pkl', 'wb') as f:
    pickle.dump(model_autoSK, f)

# load model
with open('model-classifier.pkl', 'rb') as f:
    loaded_classifier = pickle.load(f)

# predict
y_true = y_test
y_pred = loaded_classifier.predict(X_test)
print('Accuracy:', accuracy_score(y_true, y_pred))

# New prediction
y_new_pred = loaded_classifier.predict(X_new_test)
```

**Comment by eddiebergman** [comment]
Hi @jppmatos , at first glance I can't see why the data-preprocessing wouldn't be included but I may be wrong here.

Could you show exactly what component is missing or else the expected behavior versus the behavior you observe?

**Comment by jppmatos** [comment]
Thanks @eddiebergman for replying my silly question, I just wanted to be sure if the data-preprocessing is included, I've read the documentation but still wasn't sure.

I've another question: how do I select a specific pipeline model from all the found models?
Can I do something like in issue #1255, using `automodel.get_models_with_weights()`?
Like this:
```
# Save the last model:
last_weight, last_model = automodel.get_models_with_weights()[-1]

with open('iris-classifier.pkl', 'wb') as f:
      pickle.dump(last_model, f)
```

**Comment by eddiebergman** [comment]
I'll refer you to issue #1224 for a more complete walkthrough

**Comment by jppmatos** [comment]
Hello, 
`automodel.cv_results_` shows all searched models by auto-sklearn's ensembling with their following scores (like accuracy) and used parameters. But not all found models are stored? 

So that's why I can't find all the models from `automodel.cv_results` in `automodel.automl_.models_.items()` ?
Should I just increase `ensemble_size`?

**Comment by eddiebergman** [comment]
We currently don't have an easy way to access all model that autosklearn searched, only the ones that were included into the ensemble.

Increasing the `ensemble_size` may work but is not the solution, as the ensemble consists of duplicates of the same model if it increases the overall prediction performance.

The best I can direct you to is `automodel.automl_.backend.list_all_models(seed)` which returns a list of all model files. These look like `<seed>_<id>_<budget>.model`.

You can load any particular pipeline with the following
```python
backend = automodel.automl_.backend

model = backend.load_model_by_seed_and_id_and_budget(seed, id, budget)
model = backend.load_cv_model_by_seed_and_id_and_budget(seed, id, budget)
```

Id also make autosklearn not delete any temporary directories once it completes such that everything will be available.

```python
automodel = AutoSklearnClassifier(..., delete_tmp_folder_after_terminate=False)
```

We do want to change this in the future but unfortunately this is the only solution I can offer you for now.

**Comment by jppmatos** [comment]
Before asking if there was as way to get all models found in `automodel.cv_results_` I did a "hack" which finds, if there is, any models from `automodel.cv_results_` in `automodel.automl_.models_.items()` by comparing the models parameters.
```python
#...
def get_metric_result(cv_results):
	""" Makes a dataframe from cv_results_ that shows the found models and their score (accuracy, f1_score, ... ) and parameters.
	Requires: automodel.cv_results_
	Ensures: cv_results_ data in a  string and dataframe
	"""
	results = pd.DataFrame.from_dict(cv_results)
	results = results[results['status'] == "Success"]
	cols = ['rank_test_scores', 'param_classifier:__choice__', 'mean_test_score']
	cols.extend([key for key in cv_results.keys() if key.startswith('metric_')])
	#Make info_table:
	info = cols.copy()
	info.extend(('metric_balanced_accuracy','params'))
	df_infomatrix = results[info]
	#df_infomatrix = results[['rank_test_scores','param_classifier:__choice__', 'mean_test_score','metric_balanced_accuracy','params']]
	df_infomatrix.to_csv(infoTable, sep = ';' ,index_label=False) #save results report
	#
	return results[cols], df_infomatrix

def find_best_log_models(df_resultsLOG,model_autoSK,n_lim,save_path):
	"""Look for the n best models in between *.cv_results* and *.automL_.models_.items()*.
		Requires: df_resultsLOG, model.cv_results dataframe; model_autoSK, ensembled models; n_lim, the n first models from df_resultsLOG; save_path, path to save the models.
		Ensures: found models in model_autoSK.automL_.models_.items() are saved in *save_path*.
	"""
	df_bestLOG = df_resultsLOG.reset_index()
	for model_place in range(n_lim):
		selected_model_LOG_params = df_bestLOG['params'].loc[model_place]
		#selected_model:
		searchfor_model(selected_model_LOG_params,model_autoSK,save_path,model_place)

def searchfor_model(wanted_model_params,model,save_path,n_rank):
	"""Search of a model identified in models.cs_results() and save it.
	"""
	wanted_model_params = str(wanted_model_params)
	for (seed, model_id, budget), model in model.automl_.models_.items():
		models_params = model.__str__()
		search_params = models_params.split("SimpleClassificationPipeline")[1].strip("()")
		search_params = search_params.split(",\ndataset_properties")[0]
		#print(type(search_params),'\n',search_params)
		#print(search_params,'\n')
		if search_params == wanted_model_params:
			model_file = '/classifier_model_' + str(n_rank) +'_best.pkl'
			print(model_file,'\n \n',model)
			with open(save_path + model_file, 'wb') as f:
				pickle.dump(model, f)

def save_ensembled_models(models,save_path):
	""" Save models variable that has all ensemble models by auto-SKlearn.
	"""
	with open(save_path + "/all_ensembled_models.pkl", 'wb') as f:
		pickle.dump(models,f)

if __name__ == '__main__':
	automodel = AutoSklearnClassifier(time_left_for_this_task=time_run, per_run_time_limit=time_lim_per,
						n_jobs=TOTAL_JOBS,
						resampling_strategy='cv',
						resampling_strategy_arguments={'folds': n_folds},
						ensemble_size=1000,
						ensemble_nbest=200,
						max_models_on_disc=500,
						tmp_folder= temp_folder,
						delete_tmp_folder_after_terminate= True,
						scoring_functions=[balanced_accuracy,precision, recall,f1,roc_auc])

	#...

	#*.cv_results_*
	data_metrics, df_metrics_results = get_metric_result(model_autoSK.cv_results_)
	print(data_metrics.to_string(index=False),file=open(logfilename,'a'))

	#start extracting the models:
	df_metrics_results = df_metrics_results.loc[:,~df_metrics_results.columns.duplicated()].sort_values(['metric_balanced_accuracy','metric_recall'],axis=0, ascending=False)
	find_best_log_models(df_metrics_results,model_autoSK,12,models_folder) #grab models common to metrics_results
	save_ensembled_models(model_autoSK,models_folder) # save all ensemble models
	
```

In this snippet it is selected and saved the **n** best models from `automodel.cv_results_` with the higher **balanced_accuracy** and **recall**. 

However most of the models were found, but when I used another smaller dataset, none of the models were found (*"ensemble consists of duplicates..."*, how it was warned in the previous answer).


So, can I select all `automodel.automl_.models_.items()` model's and test them again, when I don't find the models from `automodel.cv_results_`. Or should I stick to the `backend` recommended approach to not lose valuable models?
(I'm just interested in the models with the higher **balanced_accuracy** and **recall**. Between 12 or 6 best models.)


Sorry for the long post/question.

**Comment by eddiebergman** [comment]
The backend should have all models there. If your goal is for models with higher `balanced_accuracy` and `recall`, I would advise using those metrics when training autosklearn as seen in this [example](https://automl.github.io/auto-sklearn/master/examples/40_advanced/example_calc_multiple_metrics.html#metrics). Of course this depends on what your end goal is but just letting you know the feature is there.

The ensemble consisting of duplicates is intended, we create the ensemble by _sampling with replacement_, adding the model if it improves the performance.

**Comment by eddiebergman** [comment]
I also updated this issue title as the question has changed quite a bit from where it started.
